### PR TITLE
groundIntake

### PR DESCRIPTION
**Pull request by MatthewChoulas**

### CODE REVIEW

File: src/main/kotlin/com/team4099/robot2023/config/constants/Constants.kt [hunk 1]
```diff
@@ -65,4 +65,10 @@ object Constants {
   object Alert {
     val TABS = arrayOf("Pre-match", "In-match")
   }
+
+  object Intake {
+    const val ROLLER_MOTOR_ID = 1
+    const val LEADER_ARM_MOTOR = 2
+    const val FOLLOWER_ARM_MOTOR = 3
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/config/constants/Constants.kt:72]: set to proper IDs

File: src/main/kotlin/com/team4099/robot2023/config/constants/IntakeConstants.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  enum class ROLLER_STATE(val power: Double) {
+    INTAKE(0.8),
+    OUTTAKE(-0.8),
+    IDLE(0.0)
+  }
+
+  val NEO_kP = 0.0.volts / 1.degrees
+  val NEO_kI = 0.0.volts / (1.degrees * 1.seconds)
+  val NEO_kD = 0.0.volts / (1.degrees.perSecond)
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/config/constants/IntakeConstants.kt:25]: change NEO_kP to "arm_kP"

File: src/main/kotlin/com/team4099/robot2023/config/constants/IntakeConstants.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  val INTAKE_ZERO = 128 // TODO: (Make this fr)
+  // From encoder to intake
+  val INTAKE_SENSOR_RATIO = (14.0 / 32.0) // TODO: (I MADE THIS UP. FIX)
+  val INTAKE_ARM_GEAR_RATIO = (24.0 / 16.0) // TODO: (I MADE THIS UP. FIX)
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/config/constants/IntakeConstants.kt:30]: figure out from design

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    followerArmMotor.enableVoltageCompensation(IntakeConstants.VOLTAGE_COMPENSATION.inVolts)
+
+    leaderArmMotor.inverted = IntakeConstants.LEFT_MOTOR_INVERTED
+    followerArmMotor.inverted = IntakeConstants.RIGHT_MOTOR_INVERTED
+
+    followerArmMotor.follow(leaderArmMotor)
+  }
+
+  override fun updateInputs(inputs: IntakeIO.IntakeIOInputs) {}
+
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt:52]: need a setAngle command

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  override fun setRollerPower(outputPower: Double) {}
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt:53]: needs to be done

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    leaderArmMotor.enableVoltageCompensation(IntakeConstants.VOLTAGE_COMPENSATION.inVolts)
+    followerArmMotor.enableVoltageCompensation(IntakeConstants.VOLTAGE_COMPENSATION.inVolts)
+
+    leaderArmMotor.inverted = IntakeConstants.LEFT_MOTOR_INVERTED
+    followerArmMotor.inverted = IntakeConstants.RIGHT_MOTOR_INVERTED
+
+    followerArmMotor.follow(leaderArmMotor)
+  }
+
+  override fun updateInputs(inputs: IntakeIO.IntakeIOInputs) {}
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt:51]: needs to be done

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIOSim.kt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.team4099.robot2023.subsystems.intake
+
+object IntakeIOSim : IntakeIO {
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIOSim.kt:3]: needs to be done

File: src/main/kotlin/com/team4099/robot2023/commands/Intake/SetAngleCommand.kt [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+ import com.ctre.phoenix.motorcontrol.ControlMode
+import com.ctre.phoenix.motorcontrol.DemandType
+import com.ctre.phoenix.motorcontrol.NeutralMode
+import com.team4099.lib.units.derived.ElectricalPotential
+import com.team4099.lib.units.derived.inVolts
+import com.team4099.lib.units.derived.volts
+import com.team4099.robot2023.config.constants.Constants
+import com.team4099.robot2023.config.constants.Constants.Universal.CANIVORE_NAME
+import com.team4099.robot2023.config.constants.IntakeConstants
+class SetAngleCommand {
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/commands/Intake/SetAngleCommand.kt:11]: oh i mistyped, i meant write setangle method not an actual command. we still need to make other commands though.

File: src/main/kotlin/com/team4099/robot2023/config/constants/Constants.kt [hunk 9]
```diff
@@ -65,4 +65,9 @@ object Constants {
   object Alert {
     val TABS = arrayOf("Pre-match", "In-match")
   }
+
+  object Led {
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/config/constants/Constants.kt:69]: This is small enough that it is probably fine but ideally we should avoid unrelated changes going in PRs -- it makes it harder to review and introduces a higher likelihood that stuff gets merged without it working

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIO.kt [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+import org.team4099.lib.units.inDegreesPerSecond
+import org.team4099.lib.units.perSecond
+
+interface IntakeIO {
+  class IntakeIOInputs : LoggableInputs {
+    var leaderArmPosition: Angle = 0.0.degrees
+    var leaderArmVelocity = 0.0.degrees.perSecond
+
+    var leaderArmAppliedOutput = 0.0
+    var followerArmAppliedOutput = 0.0
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIO.kt:21]: There is no more follower on the intake (it is one motor for pivot, one motor for rollers)

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIO.kt [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+      }
+      table?.getDouble("groundIntakeRollerTempCelcius", rollerTempCelcius.inCelsius)?.let {
+        rollerStatorCurrent = it.amps
+      }
+    }
+  }
+
+  fun updateInputs(inputs: IntakeIOInputs) {}
+
+  fun setRollerPower(outputPower: Double) {}
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIO.kt:108]: Do we want to do this or do we just want to take a voltage here? Ideally we can get rid of the percent outputs pretty much everywhere to make the behavior more predictable (so that the behavior is at least potentially the same when we are browning out vs. when the battery is new)

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+  private val leaderArmMotor =
+    CANSparkMax(Constants.Intake.LEADER_ARM_MOTOR, CANSparkMaxLowLevel.MotorType.kBrushless)
+  private val followerArmMotor =
+    CANSparkMax(Constants.Intake.FOLLOWER_ARM_MOTOR, CANSparkMaxLowLevel.MotorType.kBrushless)
+
+  private val leaderArmEncoder = leaderArmMotor.getEncoder()
+  private val followerArmEncoder = leaderArmMotor.getEncoder()
+
+  private val throughBoreEncoder = Encoder(0, 1, true, CounterBase.EncodingType.k4X)
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt:25]: Why are we using the through bore encoder as a relative encoder? The NEO's relative encoder is plenty good enough to use here. We want the through bore encoder to be an absolute encoder so that we know the position of the arm on bootup

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIO.kt [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+import org.team4099.lib.units.base.inCelsius
+import org.team4099.lib.units.derived.Angle
+import org.team4099.lib.units.derived.degrees
+import org.team4099.lib.units.derived.inDegrees
+import org.team4099.lib.units.inDegreesPerSecond
+import org.team4099.lib.units.perSecond
+
+interface IntakeIO {
+  class IntakeIOInputs : LoggableInputs {
+    var leaderArmPosition: Angle = 0.0.degrees
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIO.kt:17]: related to comment below about the absolute encoder change: we should log the absolute encoder's raw value here, along with its value multiplied by the gear ratio or whatever, so we can see over the course of a match if the NEO encoder is actually drifting

**Review by plusparth — CHANGES_REQUESTED** [review]
looks good overall, just left a couple of comments that need to be addressed

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt (hunk 12 above)
Comment (AlphaPranav9102) [reply src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIONeo.kt:25]: hmm. Did I create it wrong? The Encoder docs aren't great and I couldn't figure out how to make it an absolute encoder. What argument did I set wrong.

**Review by plusparth — CHANGES_REQUESTED** [review]
great work! I left a bunch of comments -- don't be alarmed, it is a normal number of comments for a PR of this size I think. Comments left with "nit" are nitpicks and are lower priority to fix, so I would look at the other ones first. A lot of my comments are genuine questions -- if they cause you to want to change something that's cool, but I would appreciate an answer even if you don't change anything.

**Review by plusparth — CHANGES_REQUESTED** [review]
great work! I left a bunch of comments -- don't be alarmed, it is a normal number of comments for a PR of this size I think. Comments left with "nit" are nitpicks and are lower priority to fix, so I would look at the other ones first. A lot of my comments are genuine questions -- if they cause you to want to change something that's cool, but I would appreciate an answer even if you don't change anything.

File: src/main/kotlin/com/team4099/lib/logging/TunableNumber.kt [hunk 14]
```diff
@@ -4,8 +4,8 @@ import com.team4099.robot2023.config.constants.Constants
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
 
 /**
- * Taken from
- * https://github.com/Mechanical-Advantage/SwerveDevelopment/blob/main/src/main/java/frc/robot/util/TunableNumber.java
+ * Taken from httpgit
```
[the remark this answers is not in the document]
Comment (MatthewChoulas) [reply src/main/kotlin/com/team4099/lib/logging/TunableNumber.kt:7]: fixed

File: src/main/kotlin/com/team4099/robot2023/RobotContainer.kt [hunk 15]
```diff
@@ -62,6 +88,13 @@ object RobotContainer {
     //
     // ControlBoard.advanceAndClimb.whileActiveOnce(AdvanceClimberCommand().andThen(RunClimbCommand()))
     //        ControlBoard.climbWithoutAdvance.whileActiveOnce(RunClimbCommand())
+
+    ControlBoard.extendIntake.whileTrue(IntakeExtendCommand(intake))
```
[the remark this answers is not in the document]
Comment (MatthewChoulas) [reply src/main/kotlin/com/team4099/robot2023/RobotContainer.kt]: they were mainly just for testing but the final command could be built off them. since intake command would be pretty simple idk if we'd want to do this but you could use IntakeExtendCommand in execute and then IntakeRetractCommand in interrupt of another command to create the functionality ur talking abt

File: src/main/kotlin/com/team4099/robot2023/commands/intake/IntakeCharacterizeCommand.kt [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+ * @property appliedVolts the current amount of volts being applied to the motors
+ * @property step the increase in volts per iteration
+ */
+class IntakeCharacterizeCommand(val intake: Intake) : CommandBase() {
+
+  init {
+    addRequirements(intake)
+  }
+
+  var hasMoved = false
```
[the remark this answers is not in the document]
Comment (MatthewChoulas) [reply src/main/kotlin/com/team4099/robot2023/commands/intake/IntakeCharacterizeCommand.kt]: fixed

File: src/main/kotlin/com/team4099/robot2023/commands/intake/IntakeCharacterizeCommand.kt [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+import com.team4099.robot2023.subsystems.intake.Intake
+import edu.wpi.first.wpilibj2.command.CommandBase
+import org.team4099.lib.units.derived.degrees
+import org.team4099.lib.units.derived.inVolts
+import org.team4099.lib.units.derived.volts
+import org.team4099.lib.units.perSecond
+
+/**
+ * This command is used to experimentally determine the kS value for the elevator It prints the
+ * first voltage level that causes the elevator to move
```
[the remark this answers is not in the document]
Comment (MatthewChoulas) [reply src/main/kotlin/com/team4099/robot2023/commands/intake/IntakeCharacterizeCommand.kt]: fixed

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIOSim.kt [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+    inputs.armSupplyCurrent = 0.amps
+    inputs.armTemp = 0.celsius
+
+    RoboRioSim.setVInVoltage(
+      BatterySim.calculateDefaultBatteryLoadedVoltage(armSim.currentDrawAmps)
+    )
+  }
+
+  override fun setRollerPower(voltage: ElectricalPotential) {
+    rollerSim.setInputVoltage(MathUtil.clamp(voltage.inVolts, -12.0, 12.0))
```
[the remark this answers is not in the document]
Comment (MatthewChoulas) [reply src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIOSim.kt]: I don't think there's a good reason for this so i changed it in sim

File: src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIOSim.kt [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+    inputs.armPosition = armSim.angleRads.radians
+    inputs.armVelocity = armSim.velocityRadPerSec.radians.perSecond
+    inputs.armAppliedVoltage = 0.volts
+    inputs.armStatorCurrent = armSim.currentDrawAmps.amps
+    inputs.armSupplyCurrent = 0.amps
+    inputs.armTemp = 0.celsius
+
+    RoboRioSim.setVInVoltage(
+      BatterySim.calculateDefaultBatteryLoadedVoltage(armSim.currentDrawAmps)
```
[the remark this answers is not in the document]
Comment (MatthewChoulas) [reply src/main/kotlin/com/team4099/robot2023/subsystems/intake/IntakeIOSim.kt]: yes https://team4099.slack.com/archives/C04JA631N31/p1675140991256649

**Review by plusparth — APPROVED** [review]
lgtm!
